Use standard 1px window borders on NC Island Window

greg904 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

I put the standard Windows window borders on the Windows Terminal window (can you read that?).
(It's only on the left, bottom and right though, on the top I had to use a hack because of the title bar)

Light mode, color in title bar disabled:
![frame](https://user-images.githubusercontent.com/56923875/67972211-87d2ed80-fc0e-11e9-8a89-d0a218a6fc36.png)

Light mode, color in title bar enabled:
![frame color](https://user-images.githubusercontent.com/56923875/67972645-63c3dc00-fc0f-11e9-9218-f229cdf7dfa7.png)

Dark mode, color in title bar disabled:
![frame dark](https://user-images.githubusercontent.com/56923875/67972668-6cb4ad80-fc0f-11e9-943e-3b5bd9f5079e.png)

Dark mode, color in title bar enabled:
![frame dark color](https://user-images.githubusercontent.com/56923875/67972679-72aa8e80-fc0f-11e9-9da0-8308a85359a5.png)

High contrast mode (with Settings app on right to compare):
![high contrast](https://user-images.githubusercontent.com/56923875/68075465-7b829800-fda8-11e9-8427-7069d636d666.png)

Note: the title bar should probably be purple like in Win32 apps but it isn't either in the UWP apps (Settings, classic Edge, …) so I don't really know.

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Closes #3064, closes #1307, closes #3136, closes #1897, closes #3222, closes #1859 and maybe more?
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

We take the standard window frame except that we remove the top part (see `NonClientIslandWindow::_OnNcCalcSize`), then we put little 1 pixel wide top border back in the client area using `DwmExtendFrameIntoClientArea` and then we put the XAML island and the drag bar on top.

Most of this PR is comments to explain how the code works and also removing complex code that was needed to handle the weird cases when the borders were custom.

I've also refactored a little bit the `NonClientIslandWindow` class.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

greg904 commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -531,53 +519,38 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
+                // alpha channel to 255. This is a hack to do it with GDI.
+                HDC opaqueDc;
+                BP_PAINTPARAMS params = { sizeof(params), BPPF_NOCLIP | BPPF_ERASE };
+                HPAINTBUFFER buf = BeginBufferedPaint(hdc.get(), &rcRest, BPBF_TOPDOWNDIB, &params, &opaqueDc);
```
I don't really know how GDI works so I found the first solution on Google to do this but maybe there is a better solution.

greg904 commented on src/types/lib/types.vcxproj.filters:
```diff
@@ -60,9 +60,6 @@
     <ClCompile Include="..\UTF8OutPipeReader.cpp">
       <Filter>Source Files</Filter>
     </ClCompile>
-    <ClCompile Include="..\WindowUiaProvider.cpp">
```
I don't know what this is, there is no `WindowUiaProvider.cpp` file in `src/types`. The file is in `src/interactivity/win32`. I think Visual Studio removed it automatically? Should I remove this?

DHowett-MSFT replied:
This is fine to remove. We occasionally miss updates to the `filters` files, and VS is more than happy to change them where necessary.

greg904 commented:
For now there is no resize handle directly on top of the tab row.

greg904 commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -457,47 +445,6 @@ void NonClientIslandWindow::_UpdateDragRegion()
         return 0;
     }
-
-    case WM_NCLBUTTONDOWN:
```
This is already done in `IslandWindow::MessageHandler`.

greg904 commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -457,47 +445,6 @@ void NonClientIslandWindow::_UpdateDragRegion()
-            break;
-        }
-    }
-    case WM_DPICHANGED:
```
This is already done in `BaseWindow::MessageHandler`.

miniksa commented:
I don't know enough about the non-client area and XAML stuff to go through this. @DHowett-MSFT is the non-client expert and @zadjii-msft is the current UI king. So I'll let them stab first.

greg904 commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -461,342 +428,120 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
+        // See NonClientIslandWindow::_UpdateFrameMargins for more information.
+        HDC opaqueDc;
+        BP_PAINTPARAMS params = { sizeof(params), BPPF_NOCLIP | BPPF_ERASE };
+        HPAINTBUFFER buf = BeginBufferedPaint(hdc.get(), &rcRest, BPBF_TOPDOWNDIB, &params, &opaqueDc);
```
I don't really know how GDI works so I found the first solution that worked on Google to do this but maybe there is a better solution.

greg904 commented:
Note: the big purple top border is missing when high contrast is activated on Windows because I hardcoded its size as 1 pixel * DPI scale. But it's also missing on UWP apps so I guess the people at Microsoft who did the UWP apps had the same problem but couldn't fix it so maybe there is no easy solution?

![high contrast](https://user-images.githubusercontent.com/56923875/68075428-1f1f7880-fda8-11e9-9928-8dff78f7454d.png)

greg904 commented:
Please 😢

![image](https://user-images.githubusercontent.com/56923875/68077346-53a02e00-fdc2-11e9-867a-317f0ff1c48f.png)

DHowett-MSFT commented:
This is a lot to digest, but I think I like where it's going. It definitely cleans up the reponsibilities of a bunch of the crap in NonClientIslandWindow. I'll need to give it a closer review, but it seems like a more refined version of the work I was doing to achieve the same goal. :smile:

DHowett-MSFT commented on src/cascadia/WindowsTerminal/AppHost.cpp:
```diff
@@ -188,54 +188,40 @@ void AppHost::_HandleCreateWindow(const HWND hwnd, RECT proposedRect, winrt::Ter
+        long nonClientHeight = currentHeight;
+
+        RECT frame = {};
+        bool succeeded = AdjustWindowRectExForDpi(&frame, WS_OVERLAPPEDWINDOW, false, 0, dpix);
```
(this can be a `LOG_LAST_ERROR_IF(!...`)

DHowett-MSFT commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -461,342 +424,120 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
-        }
-    }
-    else
+    if (FAILED(ThemeUtils::SetDwmImmersiveDarkMode(_window.get(), isDarkMode)))
```
this one should just be `LOG_IF_FAILED(...)` -- not because it's fewer lines of code, but because this function returns an HRESULT, and DWM returns an HRESULT, but `LOG_LAST_ERROR` uses `GetLastError()` to get the _win32 error state_ (which isn't used by any of the downstream callees here)

LOG_IF_FAILED() just takes the HRESULT return and logs that. :smile:

DHowett-MSFT replied:
It actually will shout at you if you use it wrong, but that would require us to hit a failure:

```
                WI_USAGE_ERROR_FORWARD("CALLER BUG: Macro usage error detected.  GetLastError() does not have an error.");
```

greg904 commented:
Ok so in the before last commit I fixed a bug where the theme is taken from the system settings and not the settings.json. But because the theme was not originally passed to NonClientIslandWindow, I had to change a few files like App.h, App.cpp, AppHost.cpp, etc. Please check this.

> I'd be most comfortable merging this without the dark theme fix (which I know will cause there to be white borders around the entire window instead of just at the bottom) and then bringing in the dark theme changes when the other fix has flowed back down from Windows.

I'm not sure if I understand this correctly. I rolled back `src/interactivity/win32/windowtheme.cpp` to `master` and removed the code in the function to set dark mode in ThemeUtils (now it doesn't do anything). Also I renamed the function to be more generic because apparently it's no longer a DWM window attribute to set the frame to dark theme.

> (which I know will cause there to be white borders around the entire window instead of just at the bottom)

There was no white border at the bottom, I removed the hack with `DwmExtendFrameIntoClientArea`.

Elttob commented:
One thing I've noticed - the behaviour for when title bar colours are disabled doesn't actually line up with the current/expected behaviour seen with other apps.

As an example, here's Settings while inactive and with title bar colours disabled.
Zooming in on the left border here, you can see the border is actually a transparent shade;

![image](https://user-images.githubusercontent.com/11815374/68098158-14053f00-feb3-11e9-8d96-62f2069c3f31.png)

When I focus the Settings window, the border remains transparent;

![image](https://user-images.githubusercontent.com/11815374/68098196-40b95680-feb3-11e9-8153-3703153fbc17.png)

As far as I can see, your borders are solid colour. It's a small thing, but might be best fixed before merging 🙂 ✨

DHowett-MSFT commented on src/types/ThemeUtils.cpp:
```diff
@@ -0,0 +1,19 @@
+    {
+        // TODO:GH #3425 implement the new DWM API and change
+        //  src/interactivity/win32/windowtheme.cpp to use it.
+        return S_OK;
```
Thanks for this. :smile:

DHowett-MSFT commented on src/types/precomp.h:
```diff
@@ -29,6 +29,7 @@ Module Name:
 
 // Windows Header Files:
 #include <windows.h>
+#include <dwmapi.h>
```
We probably don't need this one for the moment now that ThemeUtils is stripped, right?

greg904 commented:
It looks like it depends on the app. On my system, explorer.exe has this white border too.
![image](https://user-images.githubusercontent.com/56923875/68101110-42dbdf00-fecc-11e9-8f52-6e38af207fbe.png)

Paint has it too:
![image](https://user-images.githubusercontent.com/56923875/68101165-8afb0180-fecc-11e9-93eb-6e9f21c2293f.png)

But conhost.exe doesn't:
![image](https://user-images.githubusercontent.com/56923875/68101269-0c529400-fecd-11e9-9abc-4f156b10f935.png)

Neither does regedit:
![image](https://user-images.githubusercontent.com/56923875/68101292-255b4500-fecd-11e9-9750-30b140d247f8.png)

Explorer and Paint are two apps that modify their title bar to add additional buttons in it (explorer to open file properties and create a new folder; paint to save or go back/forward). So maybe the window border turns white automatically for some reason when you remove the title bar? Then I don't know how the Settings app does it because it has a dark border on my system but it ALSO has a custom its title bar.

![image](https://user-images.githubusercontent.com/56923875/68101404-c77b2d00-fecd-11e9-9e3e-e1026f88c21a.png)

Elttob commented:
Interesting, it's probably something to do with UWP vs non UWP apps?

greg904 commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -202,142 +254,125 @@ void NonClientIslandWindow::OnSize(const UINT width, const UINT height)
-    RECT rcWindow = BaseWindow::GetWindowRect();
+    // there isn't a SM_CYPADDEDBORDER for the Y axis
+    return ::GetSystemMetricsForDpi(SM_CXPADDEDBORDER, _currentDpi) +
+           ::GetSystemMetricsForDpi(SM_CYFRAME, _currentDpi);
```
The issue was about not using CYDRAG, not CYFRAME.

According to https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getsystemmetrics CYDRAG is:
> The number of pixels above and below a mouse-down point that the mouse pointer can move before a drag operation begins. This allows the user to click and release the mouse button easily without unintentionally starting a drag operation. If this value is negative, it is subtracted from above the mouse-down point and added below it. 

I think CYDRAG was the wrong metric. It has nothing to do with the window frame. What we want is probably CYSIZEFRAME (at least according to the doc):
>The thickness of the sizing border around the perimeter of a window that can be resized, in pixels. SM_CXSIZEFRAME is the width of the horizontal border, and SM_CYSIZEFRAME is the height of the vertical border. 
This value is the same as SM_CYFRAME.

CYSIZEFRAME is the same as CYFRAME. Maybe I should change CYFRAME to CYSIZEFRAME for more clarity?

zadjii-msft replied:
Oh good catch, I missed that. This looks like the right answer, thanks!

greg904 commented:
Ok so I found a bug where the title bar is not included in the initial height when `showTabsInTitlebar` is on so if you set `initialCols` to 10 for example it ends up being less than that (and it can even crash if you set it to 1 because the size is too small). The bug was there before the PR but I've updated a TODO in 48cd3801c3b1df22bf809314ad4696eb0441ad03 to tell what needs to be done to fix bug (because it should probably be done in another PR). Should I make a GitHub issue for it? Because the TODO says MSFT so it's not on GitHub.

@DHowett-MSFT I initially assumed that DWM would scale the borders based on the DPI but I tried changing my DPI and the top border stays 1 pixel exactly on apps like explorer.exe so I've removed the code to scale the top border according to the DPI. It will probably fix the bug with the thick borders.

greg904 commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:
```diff
@@ -202,142 +254,125 @@ void NonClientIslandWindow::OnSize(const UINT width, const UINT height)
 
-    const auto dragBarRect = GetDragAreaRect();
-    const auto nonClientHeight = dragBarRect.bottom - dragBarRect.top;
+    // At this point, we know that the cursor is inside the client area so it
```
I love deleting code 😈.

zadjii-msft commented:
@greg904 That might actually be #2061, if not it's definitely related. I'd probably just leave it as-is, and fix it as a part of #2061.

greg904 commented:
Ok, I've reverted the TODO change to leave it as-is.

DHowett-MSFT commented:
/azp run

zadjii-msft commented on src/cascadia/WindowsTerminal/AppHost.cpp:
```diff
@@ -188,54 +188,25 @@ void AppHost::_HandleCreateWindow(const HWND hwnd, RECT proposedRect, winrt::Ter
 
-        adjustedHeight = nonClient.bottom - nonClient.top;
-        adjustedWidth = nonClient.right - nonClient.left;
+        adjustedWidth = -islandFrame.left + islandWidth + islandFrame.right;
```
I'd maybe add a comment on why we don't need to do anything for the `!_useNonClientArea` case here

zadjii-msft commented on src/cascadia/WindowsTerminal/NonClientIslandWindow.h:
```diff
@@ -21,48 +21,58 @@ Author(s):
 public:
-    NonClientIslandWindow() noexcept;
+    // this is the same for all DPIs
+    static constexpr const int topBorderVisibleHeight = 1;
```
Though probably don't just commit this suggestion, that'd break things D:

musm commented:
bravo!

Stanzilla commented:
So...any way to color this new 1px border? I don't think there is a setting in Windows to color inactive window borders differently anymore, is there?

adrianghc commented:
> I'd be most comfortable merging this without the dark theme fix (which I know will cause there to be white borders around the entire window instead of just at the bottom) and then bringing in the dark theme changes when the other fix has flowed back down from Windows.

I couldn't find more details about this here, so does this mean the white border will be fixed with a new release or does the fix need to come from Windows, i.e. we would have to get 20H1 for the fix?

greg904 commented:
@Stanzilla The border behaves the same way as other windows. When it is active it will be either the accent color that you chose in Windows Settings if "Show accent color on the following surfaces -> Title bars and windows borders" is checked or a gray color otherwise. When it is inactive, it will always be gray. I don't think that there is a setting to color inactive window borders in Windows.

@adrianghc 20H1 changed the way to make the border dark. I think it is possible to detect if 20H1 is running and use the new way and if it's older than 20H1, use the old way to make the border dark so it could be fixed without waiting for 20H1 if this is done.

Stanzilla commented:
Well yeah, I don't really want to use my accent color on every window, was just wondering if there is a way to overwrite. But if that is black in 20H1 I'm okay with that

Borkason commented:
In addition to the bottom white border, I now have 3 more borders, making the issue officially 4 times worse 😂

![grafik](https://user-images.githubusercontent.com/2971735/69682595-335a4b00-10b2-11ea-94d3-3aad061dd354.png)

linux-overuser commented:
Yes indeed, the border in dark mode is still white I've noticed, which is quite bothersome to look at. 
![image](https://user-images.githubusercontent.com/18223295/69690923-4dd4fa00-109b-11ea-9d74-9de07b9f2200.png)

Bebotron commented:
Version 0.9 has rolled out and my Windows Terminal still has this problem. Any ideas as to what the status of this issue is?
![WT white borders](https://user-images.githubusercontent.com/24438203/74582276-28a7a780-4f77-11ea-8e29-ac84d903ceef.png)
![WT version](https://user-images.githubusercontent.com/24438203/74582275-280f1100-4f77-11ea-937b-3f058e4ef3dc.png)

greg904 commented:
Yes, see #3425 and maybe PR #4577.

Bebotron commented:
Thanks @greg904, so it sounds like its not a stable solution yet correct? Wondering if it might be best to sit tight a little longer until a new version of WT fixes this, or if it's worthwhile to attempt your solution in #4577

greg904 commented:
If you run the fix:
- you are running the dev or `master` version which is not stable and can have bugs
- on top of it you are running my fix which is not very tested (I made it quickly because I, too, want that bright border to go away :))

If you wait:
- [last release](https://github.com/microsoft/terminal/releases) was 2 days ago so I'm guessing the new release won't come in a few days
- maybe more stable

BTW, just to be clear:

With the fix, if you disable:
![setting](https://user-images.githubusercontent.com/56923875/74485466-1b2ce780-4ebb-11ea-8edf-f630c4e50060.png)

You get this:
![image](https://user-images.githubusercontent.com/56923875/74582965-801f3500-4fc2-11ea-9dab-dbec843ae4b1.png)

And if you enable it, you will have a border but now it will be blue:
![image](https://user-images.githubusercontent.com/56923875/74582970-9f1dc700-4fc2-11ea-9f92-6ae4ccc24a45.png)

So there still is a blue border but it is just like other apps:
![image](https://user-images.githubusercontent.com/56923875/74582978-c4123a00-4fc2-11ea-88c1-2fc0bcebf868.png)
